### ISSUE TITLE
Reduction from buffer instead of accessor.

### ISSUE THREAD

**Issue by frobnitzem** [AdaptiveCpp/AdaptiveCpp#1055]
The SYCL spec shows [constructors for reduction()](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#reduction-interface) from both a BufferT and a T-pointer, but not an accessor.

Codeplay's SYCL implements the buffer constructor, but OpenSYCL instead implements a constructor from an accessor.

The test here:
https://github.com/OpenSYCL/OpenSYCL/blob/12fdcaedfa990ab58ddf8bce304fa8cf917e6182/tests/sycl/reduction.cpp#L332

And implementation code here:
https://github.com/OpenSYCL/OpenSYCL/blob/12fdcaedfa990ab58ddf8bce304fa8cf917e6182/include/hipSYCL/sycl/libkernel/reduction.hpp#L179

should be changed to accept buffers as arguments instead of accessors.

**Comment by illuhad** [comment]
Yeah, the issue is known. The background is that we implemented reductions before the SYCL 2020 final specification was released. In the SYCL 2020 provisional spec, the interface was using accessors. That was later changed, and we have not yet made that change, because reductions are in for a larger overhaul/rewrite. The current approach does not scale well to the various backends and compilation flows we have.

Note that the reduction from USM pointers was unchanged between SYCL 2020 final and provisional, so if you use USM pointers you won't have this issue.

As far as I know Codeplay's implementation is also not complete. To my knowledge they don't support variadic (i.e. arbitrary number of) reductions in a single ranged-based parallel_for kernel, which we support.

**Comment by frobnitzem** [comment]
I made a shim that works with both compilers here: https://github.com/ornl/syclreduce
It's not super-optimized, but provides a sensible API suggestion for a "Reducer" type that
1. captures the user's identity() and combine() functions as a const class.
2. provides its own storage buffer
3. provides a get() method for recalling the result
